storage/ksql/src/main/java/io/apicurio/registry/storage/impl/ksql/KafkaSqlRegistryStorage.java

```diff
@@ -0,0 +1,518 @@
+
+/**
+ * An implementation of a registry storage that extends the basic SQL storage but federates 'write' operations
+ * to other nodes in a cluster using a Kafka topic.  As a result, all reads are performed locally but all 
+ * writes are published to a topic for consumption by all nodes.
+ * 
+ * @author eric.wittmann@gmail.com
+ */
+@ApplicationScoped
+@PersistenceExceptionLivenessApply
+@PersistenceTimeoutReadinessApply
+@Counted(name = STORAGE_OPERATION_COUNT, description = STORAGE_OPERATION_COUNT_DESC, tags = {"group=" + STORAGE_GROUP_TAG, "metric=" + STORAGE_OPERATION_COUNT})
+@ConcurrentGauge(name = STORAGE_CONCURRENT_OPERATION_COUNT, description = STORAGE_CONCURRENT_OPERATION_COUNT_DESC, tags = {"group=" + STORAGE_GROUP_TAG, "metric=" + STORAGE_CONCURRENT_OPERATION_COUNT})
+@Timed(name = STORAGE_OPERATION_TIME, description = STORAGE_OPERATION_TIME_DESC, tags = {"group=" + STORAGE_GROUP_TAG, "metric=" + STORAGE_OPERATION_TIME}, unit = MILLISECONDS)
+@Logged
+@SuppressWarnings("unchecked")
+public class KafkaSqlRegistryStorage extends AbstractSqlRegistryStorage {
+
+    private static final Logger log = LoggerFactory.getLogger(KafkaSqlRegistryStorage.class);
+
+    @Inject
+    @ConfigProperty(name = "registry.ksql.globalRuleKey", defaultValue = "__global_rule")
+    String globalRuleKey;
+    
+    @Inject
+    @ConfigProperty(name = "registry.ksql.bootstrap.servers")
+    String bootstrapServers;
+
+    @Inject
+    @ConfigProperty(name = "registry.ksql.topic", defaultValue = "ksql-journal")
+    String topic;
+
+    @Inject
+    @ConfigProperty(name = "registry.ksql.consumer.startupLag", defaultValue = "1000")
+    Integer startupLag;
+
+    @Inject
+    @ConfigProperty(name = "registry.ksql.consumer.poll.timeout", defaultValue = "1000")
+    Integer pollTimeout;
+    
+    @Inject
+    KafkaSqlCoordinator coordinator;
+
+    @Inject
+    KafkaSqlDispatcher dispatcher;
+    
+    private boolean stopped = true;
+    private KafkaProducer<String, JournalRecord> producer;
+    private KafkaConsumer<String, JournalRecord> consumer;
+    private ThreadLocal<Boolean> applying = ThreadLocal.withInitial(() -> Boolean.FALSE);
+
+    @PostConstruct
+    void onConstruct() {
+        log.info("Using Kafka-SQL storage.");
+        // Start the Kafka Consumer thread
+        consumer = createKafkaConsumer();
+        startConsumerThread(consumer);
+        
+        producer = createKafkaProducer();
+    }
+    
+    @PreDestroy
+    void onDestroy() {
+        stop();
+    }
+    
+    public void stop() {
+        stopped = true;
+    }
+
+    /**
+     * Start the KSQL Kafka consumer thread which is responsible for subscribing to the kafka topic, 
+     * consuming JournalRecord entries found on that topic, and applying those journal entries to
+     * the internal data model.
+     * @param consumer
+     */
+    private void startConsumerThread(final KafkaConsumer<String, JournalRecord> consumer) {
```

Comment:
Exactly the point in previous comment -- use existing Registry's Kafka Utils,
as then there is no need for this code + it's a lot more robust.

Reply:
Will do, thanks.